### PR TITLE
Prevent opaque types being instantiated twice with different regions within the same function

### PR DESCRIPTION
**Pull request by oli-obk**
addresses https://github.com/orgs/rust-lang/projects/22/views/1?pane=issue&itemId=41329537

r? @compiler-errors

### CODE REVIEW

File: tests/ui/impl-trait/issue-86465.rs
Comment (oli-obk) [inline tests/ui/impl-trait/issue-86465.rs:1]: This test was exactly duplicated with `fn foo` in tests/ui/type-alias-impl-trait/multiple-def-uses-in-one-fn-pass.rs (which now lives in tests/ui/type-alias-impl-trait/lifetime_mismatch.rs)

**Comment by compiler-errors** [comment]
This should probably work. Can you check that?

```rust
type Tait<'x> = impl Sized;
fn define<'a: 'b, 'b: 'a>(x: &'a u8, y: &'b u8) -> (Tait<'a>, Tait<'b>) {
    ((), ())
}
```

**Comment by compiler-errors** [comment]
@craterbot check

**Comment by oli-obk** [comment]
Almost everything in the root regressions list is spurious disk failures (and the rest is unrelated ICEs, possibly due to disk space issues), though

```
 thread 'rustc' panicked at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/compiler/rustc_serialize/src/opaque.rs:233:42:
[INFO] [stderr] range start index 1525578 out of range for slice of length 1269760
[INFO] [stderr] stack backtrace:
[INFO] [stderr]    0:     0x7fa125b6917c - std::backtrace_rs::backtrace::libunwind::trace::h1f94320f2e6a67ac
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
[INFO] [stderr]    1:     0x7fa125b6917c - std::backtrace_rs::backtrace::trace_unsynchronized::h20b345c7dcabad28
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
[INFO] [stderr]    2:     0x7fa125b6917c - std::sys_common::backtrace::_print_fmt::h1c6b579a86ac571e
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/std/src/sys_common/backtrace.rs:67:5
[INFO] [stderr]    3:     0x7fa125b6917c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h0676035e212b50ec
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/std/src/sys_common/backtrace.rs:44:22
[INFO] [stderr]    4:     0x7fa125bcb400 - core::fmt::rt::Argument::fmt::ha6041a4893633223
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/core/src/fmt/rt.rs:142:9
[INFO] [stderr]    5:     0x7fa125bcb400 - core::fmt::write::h11d7df817e44fac2
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/core/src/fmt/mod.rs:1117:17
[INFO] [stderr]    6:     0x7fa125b5d00f - std::io::Write::write_fmt::h27916ce7b82ceedf
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/std/src/io/mod.rs:1762:15
[INFO] [stderr]    7:     0x7fa125b68f64 - std::sys_common::backtrace::_print::h5c780b472386b0af
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/std/src/sys_common/backtrace.rs:47:5
[INFO] [stderr]    8:     0x7fa125b68f64 - std::sys_common::backtrace::print::h11bae2f1a0ab82fc
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/std/src/sys_common/backtrace.rs:34:9
[INFO] [stderr]    9:     0x7fa125b6bbf7 - std::panicking::default_hook::{{closure}}::h8296a8c48c99a0fb
[INFO] [stderr]   10:     0x7fa125b6b95f - std::panicking::default_hook::h99bc273f02cfb87d
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/std/src/panicking.rs:292:9
[INFO] [stderr]     Checking systemstat v0.2.3
[INFO] [stderr]   11:     0x7fa1286a1590 - std[fafc877e3f4be7b4]::panicking::update_hook::<alloc[81b46284dcdbf2c6]::boxed::Box<rustc_driver_impl[ad2e13ae3ebaaf24]::install_ice_hook::{closure#0}>>::{closure#0}
[INFO] [stderr]   12:     0x7fa125b6c338 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h17f17d8ab4e43e66
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/alloc/src/boxed.rs:2021:9
[INFO] [stderr]   13:     0x7fa125b6c338 - std::panicking::rust_panic_with_hook::h48200e0ad4451fc0
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/std/src/panicking.rs:735:13
[INFO] [stderr]   14:     0x7fa125b6c08e - std::panicking::begin_panic_handler::{{closure}}::ha531ed5b4fb897c6
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/std/src/panicking.rs:609:13
[INFO] [stderr]   15:     0x7fa125b69646 - std::sys_common::backtrace::__rust_end_short_backtrace::h934ae3f8127b5c1c
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/std/src/sys_common/backtrace.rs:170:18
[INFO] [stderr]   16:     0x7fa125b6bdf2 - rust_begin_unwind
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/std/src/panicking.rs:597:5
[INFO] [stderr]   17:     0x7fa125bc7b25 - core::panicking::panic_fmt::hd0329d80ff1c2596
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/core/src/panicking.rs:72:14
[INFO] [stderr]   18:     0x7fa125bcdd72 - core::slice::index::slice_start_index_len_fail_rt::h9029932fc1a5fcc5
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/core/src/slice/index.rs:52:5
[INFO] [stderr]   19:     0x7fa125bcdd72 - core::slice::index::slice_start_index_len_fail::h8e29ebc9dc2644ae
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/core/src/slice/index.rs:40:9
[INFO] [stderr]   20:     0x7fa12a4163bb - <rustc_metadata[e36cc7b7e19843ab]::locator::CrateLocator>::extract_one
[INFO] [stderr]   21:     0x7fa12a41505a - <rustc_metadata[e36cc7b7e19843ab]::locator::CrateLocator>::extract_lib
[INFO] [stderr]   22:     0x7fa12a30c1fa - <rustc_metadata[e36cc7b7e19843ab]::locator::CrateLocator>::maybe_load_library_crate
[INFO] [stderr]   23:     0x7fa12a30b87e - <rustc_metadata[e36cc7b7e19843ab]::creader::CrateLoader>::load
[INFO] [stderr]   24:     0x7fa12a40cf0b - <rustc_metadata[e36cc7b7e19843ab]::creader::CrateLoader>::maybe_resolve_crate
[INFO] [stderr]   25:     0x7fa129f9ccc1 - <rustc_resolve[799b7f14818798b6]::Resolver>::early_resolve_ident_in_lexical_scope
[INFO] [stderr]   26:     0x7fa129fd19d2 - <rustc_resolve[799b7f14818798b6]::Resolver>::resolve_path_with_ribs
[INFO] [stderr]   27:     0x7fa129fa53a2 - <rustc_resolve[799b7f14818798b6]::Resolver as rustc_expand[34fa88f58ed26183]::base::ResolverExpand>::resolve_imports
[INFO] [stderr]   28:     0x7fa12a1c5a1a - <rustc_expand[34fa88f58ed26183]::expand::MacroExpander>::fully_expand_fragment
[INFO] [stderr]   29:     0x7fa12a1c4738 - <rustc_expand[34fa88f58ed26183]::expand::MacroExpander>::expand_crate
[INFO] [stderr]   30:     0x7fa12a51885f - rustc_interface[d6345c2814166138]::passes::resolver_for_lowering
[INFO] [stderr]   31:     0x7fa12a517d2b - rustc_query_impl[be6bc7142f779313]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[be6bc7142f779313]::query_impl::resolver_for_lowering::dynamic_query::{closure#2}::{closure#0}, rustc_middle[fff2dae698ec3512]::query::erase::Erased<[u8; 8usize]>>
[INFO] [stderr]   32:     0x7fa12a516db3 - rustc_query_system[162c1fefa01d75b4]::query::plumbing::try_execute_query::<rustc_query_impl[be6bc7142f779313]::DynamicConfig<rustc_query_system[162c1fefa01d75b4]::query::caches::SingleCache<rustc_middle[fff2dae698ec3512]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[be6bc7142f779313]::plumbing::QueryCtxt, false>
[INFO] [stderr]   33:     0x7fa12a516449 - rustc_query_impl[be6bc7142f779313]::query_impl::resolver_for_lowering::get_query_non_incr::__rust_end_short_backtrace
[INFO] [stderr]   34:     0x7fa12a52f04b - std[fafc877e3f4be7b4]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[d6345c2814166138]::util::run_in_thread_with_globals<rustc_interface[d6345c2814166138]::interface::run_compiler<core[a27ae4b4bc8367c2]::result::Result<(), rustc_span[9fa80e483ea6db9a]::ErrorGuaranteed>, rustc_driver_impl[ad2e13ae3ebaaf24]::run_compiler::{closure#1}>::{closure#0}, core[a27ae4b4bc8367c2]::result::Result<(), rustc_span[9fa80e483ea6db9a]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a27ae4b4bc8367c2]::result::Result<(), rustc_span[9fa80e483ea6db9a]::ErrorGuaranteed>>
[INFO] [stderr]   35:     0x7fa12a52e533 - <<std[fafc877e3f4be7b4]::thread::Builder>::spawn_unchecked_<rustc_interface[d6345c2814166138]::util::run_in_thread_with_globals<rustc_interface[d6345c2814166138]::interface::run_compiler<core[a27ae4b4bc8367c2]::result::Result<(), rustc_span[9fa80e483ea6db9a]::ErrorGuaranteed>, rustc_driver_impl[ad2e13ae3ebaaf24]::run_compiler::{closure#1}>::{closure#0}, core[a27ae4b4bc8367c2]::result::Result<(), rustc_span[9fa80e483ea6db9a]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a27ae4b4bc8367c2]::result::Result<(), rustc_span[9fa80e483ea6db9a]::ErrorGuaranteed>>::{closure#1} as core[a27ae4b4bc8367c2]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
[INFO] [stderr]   36:     0x7fa125b77105 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h43cc50cdd0629acf
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/alloc/src/boxed.rs:2007:9
[INFO] [stderr]   37:     0x7fa125b77105 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h5f04d292d31a8e17
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/alloc/src/boxed.rs:2007:9
[INFO] [stderr]   38:     0x7fa125b77105 - std::sys::unix::thread::Thread::new::thread_start::h2e74a97d86ab0436
[INFO] [stderr]                                at /rustc/5f19a76e461cf77b2134117fdb950227589d6a08/library/std/src/sys/unix/thread.rs:108:17
[INFO] [stderr]   39:     0x7fa1258e1ac3 - <unknown>
[INFO] [stderr]   40:     0x7fa125972bf4 - __clone
[INFO] [stderr]   41:                0x0 - <unknown>
```

had me stumped for a moment until I realized that `opaque.rs` is not opaque types, but some encoder terminology

**Comment by compiler-errors** [comment]
I'm gonna block this on T-lang approval of the new Min-TAIT plan. Otherwise implementation looks great to me.

@rustbot blocked

**Comment by traviscross** [comment]
@rustbot labels +I-lang-nominated

Nominating for T-lang to decide whether to adopt the following rule for Rust:

> Within a single item, when multiple defining uses of an opaque type differ only by lifetime arguments, we must be able to prove that those lifetime arguments are equal, otherwise the code will be rejected, and once the new trait solver is stabilized, it will likely not be possible to lift this restriction in a fully backward compatible way.

This is called the "once modulo regions restriction".

This rule was described and proposed in the 2023-11-08 T-lang Mini-TAIT design meeting ([minutes](https://hackmd.io/CDj92gZdTzilDWORdKhLQA)): rust-lang/lang-team#233.

The effect of this rule would be to reject the following RPIT code which is currently accepted:

```rust
pub trait Trait<'x> {}
impl<'x, T> Trait<'x> for T {}

pub fn test<'a, 'b>(a: &'a (), b: &'b ()) -> impl Trait<'a> {
    // ~^ ERROR lifetime may not live long enough
    // ~| NOTE argument requires that `'b` must outlive `'a`
    if false {
        let _ = test(b, b);
    }
    a
}
```

And to reject similar code under TAIT:

```rust
#![feature(type_alias_impl_trait)]

type Tait<'x> = impl Sized;
fn test<'a, 'b>(a: &'a (), b: &'b ()) -> Tait<'a> {
    // ~^ ERROR lifetime may not live long enough
    // ~| NOTE requires that `'a` must outlive `'b`
    // ~| NOTE requires that `'b` must outlive `'a`
    if false {
        let _: Tait<'b> = b;
    }
    a
}
```

And to also reject, e.g.:

```rust
#![feature(type_alias_impl_trait)]

type Tait<'x> = impl Sized;
fn test<'a, 'b>(a: &'a (), b: &'b ()) -> (Tait<'a>, Tait<'b>) {
    // ~^ ERROR lifetime may not live long enough
    // ~| NOTE requires that `'a` must outlive `'b`
    // ~| NOTE requires that `'b` must outlive `'a`
    (a, b)
}
```

**Comment by Jules-Bertholet** [comment]
This proposal worries me. It would reject code that:

- Is sound
- Is reasonable, plausibly could have real-world application
- Intuitively "should work"
- Has no obvious work-around
- Works with the old trait solver
- And there's apparently no hope of ever making it work again, ever?

Perhaps I just need to learn more about the internals of the new solver to understand the motivation. But from an outside perspective, this does not inspire confidence.

**Comment by compiler-errors** [comment]
Let me break down this list:

* Is sound

There's lots of sound code that doesn't work in the old trait solver *or* the new trait solver.

Especially when it comes to things that are equal modulo regions, we already have a lot of limitations when it comes to the trait solver distinguishing choices (e.g. `where` clauses), because the trait solver fundamentally is not built [to distinguish lifetimes](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fb435587da1a826a40b551bd51423118).

* Is reasonable, plausibly could have real-world application

I haven't seen any examples of code that rely on this property. I agree that it's *plausible*, but it's clear that our guestimates of the probability of this being an issue that is hit in practice differ.

My understanding from discussing this briefly with @oli-obk is that he only wrote the tests that exercise this behavior because they *could* work, not because they were motivated by any real-life examples (though I may be misremembering, so any paraphrasing is my own).

* Intuitively "should work"

Again, lots of things that intuitively should work don't due to real, concrete limitations to the compiler, especially ones that enable other code to work, and for the trait solver to be sound. I go into more detail below, but this limitation exists to make space for lazy normalization and region uniquification.

* Has no obvious work-around

You can use type or const parameters to work around this, but again, a fix would likely be motivated by first understanding what is trying to be achieved at a high level.

* Works with the old trait solver

The purpose of this limitation being put in place is precisely to level out the differences between the new trait solver and the old trait solver. See below.

* And there's apparently no hope of ever making it work again, ever?

We are not technically committing to this behavior until the new trait solver lands, but yes, I don't think there's a way to make this work soundly in a language that doesn't do borrow-checking in tandem with type-checking.

Hopefully at the point of stabilization of the new solver, we will be able to more precisely motivate the reasons why the trait solver requires this to work the way it does. but TL;DR is "lazy normalization" and "region uniquification". The former is a property that we need to have to make the trait solver sound, and the latter is one that gives us better caching and is [a requirement to make the trait solver work properly](https://rustc-dev-guide.rust-lang.org/solve/invariants.html#trait-solving-must-be-free-lifetime-agnostic-) between typeck and borrowck.

The side-effect of this limitation is that we can treat defining instantiations of TAITs as being defined anywhere in the body. For example (https://godbolt.org/z/v9GjnP769) (though don't go trying it with region args yet, that still fails due to bugs with aliases in the new trait solver).

**Comment by traviscross** [comment]
@Jules-Bertholet:

> This proposal worries me....Perhaps I just need to learn more about the internals of the new solver to understand the motivation....

If you haven't yet, it's worth reading the relevant [section](https://hackmd.io/CDj92gZdTzilDWORdKhLQA#Once-modulo-regions-restriction) (and [appendix](https://hackmd.io/CDj92gZdTzilDWORdKhLQA#Appendix-C-Once-modulo-regions-restriction-details)) in the document and wrestling with the fact that we already reject this:

```rust
trait Trait<'x> {
    fn method();
}

fn test<'a, 'b, T>()
where
    T: Trait<'a> + Trait<'b>,
    // ~^ ERROR type annotations needed: cannot satisfy `T: Trait<'a>`
    // ~| NOTE multiple `impl`s or `where` clauses satisfying `T: Trait<'a>` found
{}
```

And this:

```rust
trait Trait<'a> {}
impl<T> Trait<'_> for T {}

fn capture<'a, 'b>() -> impl Sized + Trait<'a> + Trait<'b> {}

fn test<'a, 'b>() -> impl Sized + Trait<'a> + Trait<'b> {
    // ~^ ERROR lifetime may not live long enough
    // ~| HELP consider adding the following bound: `'a: 'b`
    // ~| HELP consider adding the following bound: `'b: 'a`
    // ~| HELP `'a` and `'b` must be the same: replace one with the other
    capture::<'a, 'b>()
}
```

There's no prospect of this sort of thing ever working in general in Rust without a major conceptual change to our type system.

Also, of course, see the great articulation by @compiler-errors [above](https://github.com/rust-lang/rust/pull/116935#issuecomment-1809038976).

**Comment by tmandry** [comment]
On the lang side I think we should trust the types team to move forward with this change, since it enables the new trait solver and there are no examples of breakage in crater.

Is everyone okay with making this a hard error?

@rfcbot fcp merge

However, I am hesitant about moving to a hard error _immediately_, especially given that it might result in a nontrivial refactor for someone not in crater. I'd like to make this deny-by-default and transition to a hard error some time later (maybe in the following release).

@rfcbot concern can there be a soft transition

In any case if that is hard for some reason, I wouldn't block merging this on it. I also propose that we should use this FCP for making it a hard error _eventually_ (whether in this PR or in a follow-up).

**Comment by aliemjay** [comment]
@rfcbot concern higher-ranked-lifetimes

The design document lacks specifications for the behavior of "higher ranked lifetimes," and the current implementation is concerning in this regard.

Do we have to support higher ranked lifetimes in opaque types (see #96146)? I believe the answer is yes especially given that this issue is a blocker for #117587, but then more questions arise regarding its interactions with the "once-modulo-regions" rule implemented in this PR... 

Is this code ok? I think it is despite the fact that the opaque type is instantiated with different lifetimes.
```rust
fn id(s: &str) -> &str { s }

type Opaque<'a> = impl Sized + 'a;

fn test(s: &str) -> (impl Fn(&str) -> Opaque<'_>, impl Fn(&str) -> Opaque<'_>) {
    (id, id)
}
```

Example 2; accepted[^1] by the current impl but I think it shouldn't.
```rust
fn id2<'a, 'b>(s: (&'a str, &'b str)) -> (&'a str, &'b str) { s }

type Opaque<'a> = impl Sized + 'a;

fn test(s: &str) -> impl for<'a, 'b> Fn((&'a str, &'b str)) -> (Opaque<'a>, Opaque<'b>) {
    id2
}
```

Example 3; accepted by the current impl as well.
```rust
fn id(s: &str) -> &str { s }

type Opaque<'a> = impl Sized + 'a;

fn test() -> (impl Fn(&str) -> Opaque<'_>, Opaque<'_>) {
    (id, s)
} 
```

[^1]: This still an error in the current nightly but because of #96146

File: compiler/rustc_borrowck/src/region_infer/opaque_types.rs
Comment (aliemjay) [inline compiler/rustc_borrowck/src/region_infer/opaque_types.rs:63]: ICE on `unwrap`:
```rust
fn id(s: &str) -> &str { s }
type Opaque<'a> = impl Sized + 'a;
fn test(s: &str) -> (Opaque<'_>, impl Fn(&str) -> Opaque<'_>) {
    (s, id)
}
```

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by lcnr** [comment]
checked the box for @oli-obk as they are on leave:

> Just check my box. I'll review my notifications at some point

**Comment by compiler-errors** [comment]
@craterbot check crates=https://crater-reports.s3.amazonaws.com/pr-116935/retry-regressed-list.txt

**Comment by compiler-errors** [comment]
@craterbot p=1

File: compiler/rustc_borrowck/messages.ftl
Comment (compiler-errors) [inline compiler/rustc_borrowck/messages.ftl:136]:
```suggestion
    .note = if all non-lifetime generic parameters are the same, but the lifetime parameters differ, it is not possible to differentiate the opaque types
```

**Review by compiler-errors — COMMENTED** [review]
extraneous period in diagnostics message

**Comment by compiler-errors** [comment]
Marking this as waiting-on-oli since he's away

@rustbot author

**Comment by oli-obk** [comment]
@aliemjay is your concern resolved by forbidding hkl and requiring anything enabling them to go through a separate feature gate? I added more tests in https://github.com/rust-lang/rust/pull/121386

**Comment by oli-obk** [comment]
> In any case if that is hard for some reason, I wouldn't block merging this on it. I also propose that we should use this FCP for making it a hard error eventually (whether in this PR or in a follow-up).

@tmandry I don't think it is hard, just undesirable to go with a future incompat lint if we find it unlikely to break user code anywhere. It requires some very interesting recursive RPIT code in order to hit any issue here at all, and while people do use RPIT with recursion, so far we have not seen anyone trying to be tricky with swapping lifetimes across the invocations *and* using that to constrain a hidden type via the returned value.

**Comment by traviscross** [comment]
@rustbot labels +I-lang-nominated

Let's nominate as this is blocking #120700 and has open concerns.

**Comment by tmandry** [comment]
@rfcbot resolve can there be a soft transition

**Comment by nikomatsakis** [comment]
@aliemjay I'd like to repeat your concern back to you to make sure that I am following:

Example 1, which you claimed should be ok...

```rust
fn id(s: &str) -> &str { s }

type Opaque<'a> = impl Sized + 'a;

fn test(s: &str) -> (impl Fn(&str) -> Opaque<'_>, impl Fn(&str) -> Opaque<'_>) {
    (id, id)
}
```

I think the idea here is that, when we go to prove that `id <: impl for<'x> Fn(&'x str) -> Opaque<'x>`, we will wind up with only one definition of `Opaque` per tuple-element. In other words, when we open the `for<'x>`-quantifier, we only do one unification of `Opaque` in that scope. I'm struggling to put this into proper words =) I guess it has to do with the proof tree, and the idea that you don't have two "sibling" unifications for distinct regions within a given quantification scope.

In contrast, with example 2, currently accepted but you think it should not be:

```rust
fn id2<'a, 'b>(s: (&'a str, &'b str)) -> (&'a str, &'b str) { s }

type Opaque<'a> = impl Sized + 'a;

fn test() -> impl for<'a, 'b> Fn((&'a str, &'b str)) -> (Opaque<'a>, Opaque<'b>) {
    id2
}
```

Here, we wind up unifying `Opaque`  twice within the same `for`-scope, right?

Can you (or someone) share a bit about how this is implemented and why it accepts what it accepts?

**Comment by oli-obk** [comment]
> In contrast, with example 2, currently accepted but you think it should not be:

we're rejecting it now: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=c4a2afe7d563825bcd5422821eaf0503

**Comment by nikomatsakis** [comment]
@oli-obk in @aliemjay's original post, they noted that it was being rejected, but not for the reason they thought it should be. Has this been updated since?

**Comment by oli-obk** [comment]
No, there was no update, except some tests ensuring we don't accidentally start accepting it: https://github.com/rust-lang/rust/pull/121386

**Comment by nikomatsakis** [comment]
@oli-obk ok, so the idea is that we punted on those details for now?

**Comment by oli-obk** [comment]
Yes, there are too many questions about higher kinded things in opaques, that we want to develop them under a separate feature gate.

**Comment by aliemjay** [comment]
While the concern of higher-ranked-lifetimes is still applicable and may block the new capture rules (#117587), I don't see a reason for it to block ATPIT stabilization since we have a hard error now when opaque types mention higher ranked regions (#121386) and I believe that's forward compatible with any future decision. For this reason I am going to resolve my concern here and mention it as a blocker in #117587. Thanks to @traviscross for bringing this to my attention. 

@rfcbot resolve higher-ranked-lifetimes
@rfcbot reviewed

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**
